olp-cpp-sdk-core/src/cache/DiskCache.cpp

```diff
@@ -384,34 +384,44 @@ DiskCache::OperationOutcome DiskCache::ApplyBatch(
 }
 
 bool DiskCache::RemoveKeysWithPrefix(const std::string& prefix,
-                                     uint64_t& removed_data_size) {
+                                     uint64_t& removed_data_size,
+                                     const RemoveFilterFunc& filter) {
   uint64_t data_size = 0u;
-  if (!database_) {
-    removed_data_size = 0u;
-    OLP_SDK_LOG_ERROR(kLogTag,
-                      "RemoveKeysWithPrefix: Database is uninitialized");
-    return false;
-  }
-
-  auto batch = std::make_unique<leveldb::WriteBatch>();
+  removed_data_size = 0u;
 
+  // As we remove data probably it is not wise to flood the leveldb memory cache
   leveldb::ReadOptions opts;
   opts.verify_checksums = check_crc_;
-  opts.fill_cache = true;
+  opts.fill_cache = false;
   auto iterator = NewIterator(opts);
+  if (!iterator) {
+    OLP_SDK_LOG_WARNING(kLogTag,
+                        "RemoveKeysWithPrefix: Database is uninitialized");
+    return false;
+  }
+
+  auto batch = std::make_unique<leveldb::WriteBatch>();
 
   if (prefix.empty()) {
-    for (iterator->SeekToFirst(); iterator->Valid(); iterator->Next()) {
-      batch->Delete(iterator->key());
-      data_size += iterator->value().size() + iterator->key().size();
-    }
+    iterator->SeekToFirst();
   } else {
-    for (iterator->Seek(prefix);
-         iterator->Valid() && iterator->key().starts_with(prefix);
-         iterator->Next()) {
-      batch->Delete(iterator->key());
-      data_size += iterator->value().size() + iterator->key().size();
+    iterator->Seek(prefix);
+  }
+
+  auto condition = [&](const leveldb::Iterator& it) {
+    return it.Valid() && (prefix.empty() || it.key().starts_with(prefix));
+  };
+
+  for (; condition(*iterator); iterator->Next()) {
+    auto key = iterator->key();
+
+    // Do not delete if protected
+    if (filter && filter(key.ToString())) {
```

Comment:
I would prefer to use string view here instead of copy on each call (yes, we need to port it). Data and size are available in the Slice so should be pretty straightforward.

Reply:
I don't think we need it as it is happening on remove either way and it is accepted to take long.

Reply:
Sadly it is currently not possible either way as the protected list uses `std::set::lower_bound()` which in C++11 expects a string as an input so no chance to avoid copying. I tried locally to pass `leveldb::Slice` instead-but in the end it needs conversion to `std::string`. The only possibility is to refactor the ProtectedKeyList, which in my opinion can be postponed for later in case there are some real performance issues, which may bother some of our users.

Reply:
COW is no longer a go-to solution for compilers for copying strings as far as I know. And for SSO those strings are too long.
So yeah, let's see the results on the real data.